app/src/behavior_key_scanned.c

```diff
@@ -0,0 +1,194 @@
+    const struct behavior_key_scanned_config *config = dev->config;
+
+    gpio_pin_interrupt_configure_dt(&config->key, active_scanning ? GPIO_INT_EDGE_TO_ACTIVE
+                                                                  : GPIO_INT_LEVEL_ACTIVE);
+}
+
+static void bks_disable_interrupt(const struct device *dev) {
+    const struct behavior_key_scanned_config *config = dev->config;
+
+    gpio_pin_interrupt_configure_dt(&config->key, GPIO_INT_DISABLE);
+}
+
+static void bks_read(const struct device *dev) {
+    const struct behavior_key_scanned_config *config = dev->config;
+    struct behavior_key_scanned_data *data = dev->data;
+
+    if (k_sem_take(&data->sem, K_NO_WAIT) < 0) {
+        // k_work_reschedule(&data->update_work, K_NO_WAIT);
+        return;
+    }
+
+    zmk_debounce_update(&data->debounce_state, data->active_scan_detected,
+                        config->debounce_scan_period_ms, &config->debounce_config);
+
+    if (zmk_debounce_get_changed(&data->debounce_state)) {
+        const bool pressed = zmk_debounce_is_pressed(&data->debounce_state);
+
+        struct zmk_behavior_binding_event event = {.position = INT32_MAX,
+                                                   .timestamp = k_uptime_get()};
+
+        if (pressed) {
+            behavior_keymap_binding_pressed(&data->binding, event);
+        } else {
+            behavior_keymap_binding_released(&data->binding, event);
+        }
+    }
+
+    if (zmk_debounce_is_active(&data->debounce_state)) {
+        data->active_scan_detected = false;
+        data->read_time += config->debounce_scan_period_ms;
+
+        k_work_schedule(&data->update_work, K_TIMEOUT_ABS_MS(data->read_time));
+    } else {
+        bks_enable_interrupt(dev, false);
+    }
+
+    k_sem_give(&data->sem);
+}
+
+static void bks_update_work(struct k_work *work) {
+    struct k_work_delayable *dwork = CONTAINER_OF(work, struct k_work_delayable, work);
+    struct behavior_key_scanned_data *data =
+        CONTAINER_OF(dwork, struct behavior_key_scanned_data, update_work);
+    bks_read(data->dev);
+}
+
+static void bks_gpio_irq_callback(const struct device *port, struct gpio_callback *cb,
+                                  const gpio_port_pins_t pin) {
+    struct behavior_key_scanned_data *data =
+        CONTAINER_OF(cb, struct behavior_key_scanned_data, key_callback);
+    const struct behavior_key_scanned_config *config = data->dev->config;
+
+    uint32_t time = k_uptime_get();
+
+    if (k_sem_take(&data->sem, K_MSEC(10)) < 0) {
+        LOG_ERR("FAILED TO TAKE THE SEMAPHORE");
+        // Do more?
+        return;
+    }
+
+    data->active_scan_detected = true;
+    data->read_time = time;
+
+    if (!zmk_debounce_is_active(&data->debounce_state)) {
+        // When we get that very first interrupt, we need to schedule the update checks to fall in
+        // between each of the real scans, so we can do our checks for state *after* each scan has
+        // occurred.
```

Comment:
This is trying to schedule this device's reads to happen in between the kscan driver's scans so they don't conflict? If so, this seems really fragile to me. Off the top of my head, here are some things that could break it:

- The scan periods are set to different things on the two devices.
- You're using a bunch of GPIO expanders or something, and doing one scan takes longer than half the scan period.

Reply:
Yeah, so this is the best I can imagine for handling this, which is the best way I can think of to detect this.

The scan periods need to be at least the matrix scan rate *or larger* and it should be ok. Definitely a risk, but not one I can easily see a workaround for.

As for slower scanning... I tweaked the scheduling to be just 1ms before the next debounce scanning period, to reduce the risk of the second scenario happening.